### PR TITLE
Use priority queue to manage proof goals

### PR DESCRIPTION
**Pull request by makaimann**
Use a priority queue for managing proof goals -- which will always return a proof goal at the lowest frame available. This PR also fixes an issue where proof goals were popped too early. Instead, they should only be removed from the queue when it has been blocked.

### CODE REVIEW

**Comment by makaimann** [comment]
Marked `do not merge` for now while we're evaluating it.

**Comment by makaimann** [comment]
FYI @zhanghongce, once this PR is merged it will change the interface slightly -- it uses `ProofGoal *` instead of just `ProofGoal`

File: engines/ic3base.cpp
Comment (ahmed-irfan) [inline engines/ic3base.cpp:135]: where do we  delete ?

File: engines/ic3base.cpp
Comment (ahmed-irfan) [inline engines/ic3base.cpp:305]: ditto

File: engines/ic3base.cpp
Comment (ahmed-irfan) [inline engines/ic3base.cpp:402]: are we going to change this later? if not then it should be a pointer to const object

**Comment by ahmed-irfan** [comment]
left some comments

File: engines/ic3base.cpp
Comment (makaimann) [reply engines/ic3base.cpp]: Oops, good catch. Thanks! Fixed with 5c927cdc9fbcad0b3f270f0ba117cd447278a72a.

File: engines/ic3base.cpp
Comment (makaimann) [reply engines/ic3base.cpp:316]: Thanks! Fixed with 5c927cdc9fbcad0b3f270f0ba117cd447278a72a

File: engines/ic3base.cpp
Comment (makaimann) [reply engines/ic3base.cpp]: Good point! Updated with 9f493ef01f129dbc798555683b27a762a3bf5b98

**Comment by makaimann** [comment]
Looks like performance actually got worse. It still seems like this is the right thing to do though. What do you think @ahmed-irfan?

Screenshot from 2020-12-29 10-03-36

**Review by ahmed-irfan — APPROVED** [review]
LGTM.

I think that we are doing the right changes in this PR. We should merge it.

**Pull request merged**
